Add Aimv2 model

yaswanth19 opened the pull request:
# What does this PR do?

Fixes #35351

yaswanth19 committed:
Model skelton

yaswanth19 committed:
changes

yaswanth19 committed:
temp push

yaswanth19 committed:
changes

yaswanth19 committed:
Added support for aimv2-native

yaswanth19 committed:
More changes

yaswanth19 commented:
@qubvel I am working on this PR which uses Modular and more adhering to HF code standards. The AIMv2 model has three variants:
1.) Normal image feature extraction (resized images to standard size): apple/aimv2-large-patch14-224 
2.) Image feature extraction checkpoint for native image resolution: apple/aimv2-large-patch14-native
3.) Vision + Text backbone model similar to clip: apple/aimv2-large-patch14-224-lit

I’ve implemented my code to support the first two model variants for image feature extraction, as there doesn’t seem to be much traction for Variant 3 based on HF Hub downloads.

However, one thing I couldn’t understand from the codebase is the use of a mask in their codebase. Since we’re using the CLIP processor, it doesn’t return a mask when an image is passed. Could you clarify why a mask is required in this context? and if required how can we generate one 🤔

yaswanth19 committed:
More changes

yaswanth19 commented:
@qubvel or @amyeroberts A gentle ping to clear my query so that I can continue the work 😅

yaswanth19 commented:
Hi @qubvel , A gentle ping to address my above query.

yaswanth19 commented:
@qubvel or @yonigozlan , A gentle ping 😅 to address my above query

yonigozlan commented:
Hey @yaswanth19 , sorry for the delay, having a look now!

yonigozlan commented:
It does look like the attention_mask (`mask`) in their code is not used, event in the "lit" variant. It's kind of similar to CLIP though, where the CLIP vision encoder (CLIPEncoder) also supports an attention mask, however there's no way to pass it in starting at CLIPVisionTransformer. I wouldn't worry too much about it, it could still be nice to have it in AIMv2VisionModel for experimentation, but it doesn't look like there's anything for you to add here.

As for the PR itself, if we are to add Aimv2 to Transformers, I'd say it would still be nice to also have support for the "lit" variant to be more comprehensive. Also it doesn't look like there are official checkpoints from apple for image classification is there? So it might not make sense to add official support for an Aimv2ForImageClassification, at least in this state.

yaswanth19 commented:
@yonigozlan Thank you for clarifying, then ig we can have vision+text as AIMv2 Model and users can load vision model to load feature extraction ckpts. So, If I am not wrong then we have to create two different conversion files one for `lit variant` which would be `AIMv2Model` and other one for feature extarction ckpts which will be loading in `AIMv2VisionModel`.

yonigozlan commented:
@yaswanth19 You could have only one conversion file taking the original checkpoint as an argument, and load/save the correct model depending on what is in the original config

yaswanth19 committed:
Stupid mistake correction

yaswanth19 committed:
Added config and refactor

yaswanth19 committed:
Added vison model

yaswanth19 committed:
update

yaswanth19 committed:
Refactor for lit variant

yaswanth19 committed:
Added Text Model

yaswanth19 committed:
Minor fixes

yaswanth19 committed:
nits

yaswanth19 committed:
update

yaswanth19 committed:
Preliminary tests

yaswanth19 committed:
More fixes

yaswanth19 committed:
Updated tests 🤗

yaswanth19 committed:
Refactor

yaswanth19 committed:
Updated testcase

yaswanth19 committed:
Updated config

yaswanth19 committed:
make fixup

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
more fixes

yaswanth19 committed:
Bug fix and updates

yaswanth19 committed:
deadcode

yaswanth19 committed:
Fixes

yaswanth19 committed:
nit

yaswanth19 committed:
up

yaswanth19 committed:
Happy CI ✅

yaswanth19 committed:
Reduce LOC

yaswanth19 committed:
nit

yaswanth19 committed:
nit

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
make style

yaswanth19 commented:
@yonigozlan  The PR is in good shape and should be ready for review soon. At the end I am stuck at one place that is the tests `test_attention_outputs` and `test_retain_grad_hidden_states_attentions` are failing locally. I am using new attention interface and by default it use sdpa which doesn't return any attn_weghts hence None is returned. Because of that these tests are failing. I have tried with eager implementation and all those tests are passing and even in test cases I don't see anywhere config._attn_implementation='eager` being set so I not sure what mistake I am making or how the other models are handling it. Can you help me in debugging this testcase?

yaswanth19 committed:
return_dict refactor

yaswanth19 committed:
bug fix

yaswanth19 committed:
fix

yaswanth19 committed:
doc update

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
@qubvel @yonigozlan The PR is ready for initial review 🤗 . 

Note: This whole arch was trained in a autoregressive way and the authors have not open sourced the decoder weights hence this model version is strictly for inference.

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
nit

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
make fixup

yaswanth19 committed:
Minor update

yaswanth19 commented:
@qubvel A gentle reminder for the initial review. It would be great to get this integrated into `transformers`, as it’s a solid encoder model.

As an example, [`AIDC-AI/Ovis2-2B`](https://huggingface.co/AIDC-AI/Ovis2-2B/tree/main) combines Qwen2VL with Apple’s AIMv2 ViT, showing that this model is already getting some traction.

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
@yonigozlan @qubvel A gentle ping for review whenever you have the bandwidth 🤗

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
_init_weigths modifcation

yaswanth19 commented:
Yup NP, one test is failing because of _init_weights refactor by Cyril, I am not sure how to handle that as that test is only run for recent and specific models.

yaswanth19 commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+            module.weight.data.fill_(1.0)
+        elif isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+        elif hasattr(module, "logit_scale"):
+            if isinstance(module.logit_scale, nn.Parameter):
+                module.logit_scale.data.fill_(math.log(1 / 0.07))
```
After the latest refactor in #37341 - the introduced test is failing. I have manged to workaround for logit_scale but we it's failing for cls_token in `AIMv2AttentionPoolingHead`. Don't know how to initilialize parameter without iterating  on all parameters or doing recursive check.

qubvel replied:
hey @yaswanth19 

You can add smth like:

```py
if isinstance(module, AIMv2AttentionPoolingHead):
    module.cls_token..  # do init here
```

yaswanth19 commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+                module.logit_scale.data.fill_(math.log(1 / 0.07))
+
+
+class AIMv2VisionModel(AIMv2PreTrainedModel):
+    main_input_name = "pixel_values"
+
+    def __init__(self, config: AIMv2VisionConfig):
+        super().__init__(config)
+        self.config = config
+        self.embeddings = AIMv2VisionEmbeddings(config)
+        self.encoder = AIMv2Encoder(config)
+        self.rms_norm = AIMv2RMSNorm(config.hidden_size, config.rms_norm_eps)
+
+        self.use_head = config.use_head
+        if self.use_head:
+            self.head = AIMv2AttentionPoolingHead(config)
+
+        self.post_init()
```
This block is similar to `SiglipVisionTransformer` and ideally we need to only overwrite the init block and use modular but the catch is we are using rms_norm here whereas in siglip its layernorm, so for correct naming convention I have overwritten everything

qubvel replied:
Can you add a comment please, smth like

```py
# The only change from SigLIP model, layernorm -> rms_norm
self.rms_norm = AIMv2RMSNorm(config.hidden_size, config.rms_norm_eps)
```

yaswanth19 commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+        input_ids: Optional[torch.LongTensor] = None,
+        pixel_values: Optional[torch.FloatTensor] = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        return_loss: Optional[bool] = None,
```
Not using this arg - as we don't have loss and we don't support training. Some tests were failing when I removed this arg as forward expects this. I can remove this and add kwargs so the argument is passed but not used. WDYT?

yonigozlan replied:
As I said above, I think we can leave it, even if loss will always be None, for consistency within the library

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+        hidden_states = self.patch_embed(pixel_values).flatten(2).transpose(1, 2)
+        hidden_states = self.rms_norm(hidden_states)
+
+        if self.config.image_size != height or self.config.image_size != width:
+            pos_embed = self.build_2d_sincos_position_embedding(
+                height // self.patch_size,
+                width // self.patch_size,
+                embed_dim=self.config.hidden_size,
+                device=hidden_states.device,
+                dtype=hidden_states.dtype,
+            )
+        else:
+            pos_embed = self.position_embedding(self.position_ids)
```
I don't think this is enough to differentiate Aimv2 native and AimV2 for fixed resolutions. The embeddings strategy should depend on the model config itself, and not the pixel values inputs. Maybe we could add something like is_native in the config?

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+        )
+        self.rms_norm = AIMv2RMSNorm(config.hidden_size, config.rms_norm_eps)
+
+        num_patches = (config.image_size // config.patch_size) ** 2
+        self.position_embedding = nn.Embedding(num_patches, config.hidden_size)
```
Same here, self.position_embedding shouldn't exist with the native model, so we would get a missing weights warning here which we don't want. Let's define self.position_embedding or not based on the config

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,804 @@
+        self.text_projection = nn.Linear(self.text_embed_dim, self.projection_dim, bias=False)
+
+        self.logit_scale = nn.Parameter(torch.tensor(self.config.logit_scale_init_value))
+        self.max_logit_scale = math.log(config.max_logit_scale)
```
Might makes sense to keep the name max_log_logit_scale given the definition

yonigozlan commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,685 @@
+        parent,
+        batch_size=12,
+        seq_length=7,
+        is_training=True,
```
```suggestion
        is_training=False,
```

yonigozlan commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,685 @@
+        image_size=30,
+        patch_size=2,
+        num_channels=3,
+        is_training=True,
```
```suggestion
        is_training=False,
```

yonigozlan commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,685 @@
+
+
+class AIMv2ModelTester:
+    def __init__(self, parent, text_kwargs=None, vision_kwargs=None, is_training=True):
```
```suggestion
    def __init__(self, parent, text_kwargs=None, vision_kwargs=None, is_training=False):
```

yonigozlan commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,685 @@
+        [ 2.7972, -2.3180, -0.4490,  2.2302],
+        [ 2.8584, -2.5322, -0.2302,  2.4936],
+        [-2.7849,  2.4121,  1.3670, -1.5514]]).to(model.device)
+        # fmt: on
+
+        output_slice = output.last_hidden_state.squeeze(0)[0:7, 0:4]
+        self.assertTrue(torch.allclose(output_slice, expected_logits, atol=1e-3))
+
+    @slow
+    def test_inference_for_native_resolution(self):
+        model_name = "yaswanthgali/aimv2-large-patch14-native-HF"
+
+        model = AIMv2VisionModel.from_pretrained(model_name, device_map="auto")
+        processor = AutoImageProcessor.from_pretrained(model_name)
+
+        image = image = Image.open(
+            requests.get("http://images.cocodataset.org/val2017/000000039769.jpg", stream=True).raw
+        )
+        inputs = processor(image, return_tensors="pt").to(model.device)
+
+        with torch.no_grad():
+            output = model(**inputs)
+
+        # Verify logits shape
+        self.assertEqual(output.last_hidden_state.shape, torch.Size([1, 1530, 1024]))
+
+        # Verify logits slice
+        # fmt: off
+        expected_logits = torch.tensor(
+        [[-1.3342,  0.3720,  0.0963,  0.4159],
+        [-1.5328,  0.4677,  0.0936,  0.4321],
+        [-0.3775, -0.2758, -0.0803, -0.5367],
+        [-1.3877,  0.5561, -1.9064, -1.1766],
+        [-0.5148,  0.0108, -0.4515, -0.6402],
+        [-0.3400, -0.1711, -0.1855, -0.4219],
+        [-1.2877, -0.0585, -0.1646,  0.7420]]).to(model.device)
+        # fmt: on
+
+        output_slice = output.last_hidden_state.squeeze(0)[0:7, 0:4]
+        self.assertTrue(torch.allclose(output_slice, expected_logits, atol=1e-3))
```
Nice!

yonigozlan commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,685 @@
+        config_and_inputs = self.model_tester.prepare_config_and_inputs()
+        self.model_tester.create_and_check_model(*config_and_inputs)
+
+    @unittest.skip
+    def test_training(self):
+        pass
+
+    @unittest.skip
+    def test_training_gradient_checkpointing(self):
+        pass
+
+    @unittest.skip(reason="This model has no Loss")
+    def test_training_gradient_checkpointing_use_reentrant(self):
+        pass
+
+    @unittest.skip(reason="This model has no Loss")
+    def test_training_gradient_checkpointing_use_reentrant_false(self):
+        pass
```
to remove after setting is_training to False

yonigozlan reviewed (commented):
Very nice work @yaswanth19 ! Much more "Transformers-like" than the original implementation, love it :). I raised a concern on the difference in `pos_embed` between the native and fixed resolution models. As there are distinct checkpoints for native and fixed resolution, I think the difference should be set at the config level.
Also you are missing the `AutoProcessor` and `AutoImageProcessor` that you should set to `CLIPProcessor` and `CLIPImageProcessor` in processing_auto.py and image_processing_auto.py

yaswanth19 commented on docs/source/en/model_doc/aimv2.md:
````diff
@@ -0,0 +1,115 @@
+)
+outputs = model(**inputs)
+probs = outputs.logits_per_image.softmax(dim=-1)
+```
````
We do have one for `native` above. For `base` it would be same and we will handle it in config implicitly so ig a separate example is not needed.

yonigozlan replied:
Missed it sorry! And ok makes sense for `base`

yaswanth19 committed:
update tests

yaswanth19 committed:
Minor fixes post review

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Update w.r.t GradientCheckpointingLayer

yaswanth19 commented:
@yonigozlan Addressed the comments. Ready for review again!!

yaswanth19 committed:
docs update

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -486,7 +463,7 @@ def forward(
 
 
-class AIMv2EncoderLayer(nn.Module):
+class AIMv2EncoderLayer(GradientCheckpointingLayer):
```
Thanks for updating

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -551,6 +528,37 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+    Args:
+        pixel_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Pixel values. Padding will be ignored by default should you provide it. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`AIMv2ImageProcessor.__call__`] for details.
```
```suggestion
            [`AutoImageProcessor`]. See [`CLIPImageProcessor.__call__`] for details.
```

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -551,6 +528,37 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
         return output
 
 
+AIMV2_START_DOCSTRING = r"""
```
We should also have a `AIMV2_VISION_START_DOCSTRING`

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -551,6 +528,37 @@ def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+    and behavior.
+
+    Parameters:
+        config ([`AIMv2Config`]): Model configuration class with all the parameters of the model.
```
With `AIMv2VisionConfig` here for `AIMV2_VISION_START_DOCSTRING`

yonigozlan commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -716,11 +755,10 @@ def __init__(self, config: AIMv2Config):
 
         self.post_init()
 
-    @can_return_tuple
```
Why was this removed?

yaswanth19 replied:
Now we are inheriting all the decorators for AIMv2Model forward class from CLIP model itself. (can_return_tuple + other doc string decorators)

yonigozlan replied:
i see thanks!

yonigozlan commented on src/transformers/models/auto/processing_auto.py:
```diff
@@ -45,6 +45,8 @@
 PROCESSOR_MAPPING_NAMES = OrderedDict(
     [
+        ("aimv2", "CLIPProcessor"),
+        ("aimv2_vision_model", "CLIPProcessor"),
```
No need for an AutoProcessor for a vision only model
```suggestion
```

yaswanth19 replied:
ahh my bad😅

yonigozlan reviewed (commented):
Thanks for iterating @yaswanth19 ! Only a few things left to change for me, then let's ask a final review from a core maintainer. We'll also have to see where we want to put the transformers native checkpoints.

yaswanth19 committed:
update

yaswanth19 committed:
nit

yaswanth19 commented on src/transformers/models/aimv2/modeling_aimv2.py:
```diff
@@ -0,0 +1,985 @@
+        >>> import requests
+        >>> from transformers import AutoProcessor, AIMv2Model
+
+        >>> model = AIMv2Model.from_pretrained("openai/aimv2-vit-base-patch32")
+        >>> processor = AutoProcessor.from_pretrained("openai/aimv2-vit-base-patch32")
```
One issue of Modular 😅

yonigozlan replied:
Yes indeed, @Cyrilvallez , I think we could try to replace the checkpoint using the one defined in the config docstring, that's what I do in auto_docstring :). I'll try to open a PR to see if we can fix that.
For now I don't think it's that important so let's not worry about it for this model

yonigozlan reviewed (commented):
Thanks for making the changes @yaswanth19! Looks good to me. Pinging @Cyrilvallez @ArthurZucker for a final review!

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+logger = logging.get_logger(__name__)
+
+
+class AIMv2VisionConfig(SiglipVisionConfig):
```
Would be nice to have `AimV2` or ` Aimv2`!

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+    pass
+
+
+class AIMv2SwiGLUFFN(nn.Module):
```
this already exists in quite a lot of models! We just call it MLP 🤗 
```python
class Llama4TextMLP(nn.Module):
    def __init__(self, config, intermediate_size=None):
        super().__init__()

        if intermediate_size is None:
            intermediate_size = config.intermediate_size

        self.config = config
        self.gate_proj = nn.Linear(config.hidden_size, intermediate_size, bias=False)
        self.up_proj = nn.Linear(config.hidden_size, intermediate_size, bias=False)
        self.down_proj = nn.Linear(intermediate_size, config.hidden_size, bias=False)
        self.activation_fn = ACT2FN[config.hidden_act]

    def forward(self, x):
        down_proj = self.activation_fn(self.gate_proj(x)) * self.up_proj(x)
        return self.down_proj(down_proj)
``` 
I think we have one that uses config.bias already! Can you check?

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+        hidden_states = self.patch_embed(pixel_values).flatten(2).transpose(1, 2)
+        hidden_states = self.rms_norm(hidden_states)
+
+        if self.config.is_native:
```
is this used in released checkpoints?

yaswanth19 replied:
Yup, we has two checkpoints:

1.)  `apple/aimv2-large-patch14-224` : This resizes the image standard dim.
2.)  `apple/aimv2-large-patch14-native`: This can be used on top of native resolution of image. To differentiate the between these two checkpoints I have used this arg

Refer: https://github.com/huggingface/transformers/pull/36625#discussion_r2054319568

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+        self.k_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=config.qkv_bias)
+        self.v_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=config.qkv_bias)
+        self.q_proj = nn.Linear(self.embed_dim, self.embed_dim, bias=config.qkv_bias)
+        self.proj_out = nn.Linear(self.embed_dim, self.embed_dim, bias=config.qkv_bias)
+        self.proj_drop = nn.Dropout(config.projection_dropout)
+
```
why not call these `o_proj` as our conventions!

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+    pass
+
+
+class AIMv2Attention(nn.Module):
```
AFAIK this already exists in transformers so we can import it (prob siglip or Clip no?)

yaswanth19 replied:
Yes, we only had a diff of projection dropout and since no checkpoint was using it - I have removed it and now we can use siglip attention 🤗

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+        # The only change from SiglipVisionTransformer is, layernorm -> rms_norm.
+        self.rms_norm = AIMv2RMSNorm(config.hidden_size, config.rms_norm_eps)
+
+        self.use_head = config.use_head
+        if self.use_head:
+            self.head = AIMv2AttentionPoolingHead(config)
```
not a fan of having codepathes, is this used for released checkpoints? 🤗

yaswanth19 replied:
Yes, we have Multiple different checkpoints and the Image feature extraction (vit like) ckpts don't have the head. The head is used by one checkpoints which will be loaded into Aimv2model as it can handle both images and text (clip like)

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+        )
+
+
+class AIMv2Model(CLIPModel, nn.Module):
```
AltCLIPModel or CLIPModel's forward and init should be more reused is here no?

yaswanth19 replied:
we can't use other clip vairants coz they don't have `_get_vector_norm`

ArthurZucker commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,844 @@
+        image_embeds = image_embeds / _get_vector_norm(image_embeds)
+        text_embeds = text_embeds / _get_vector_norm(text_embeds)
+
+        logit_scale = self.logit_scale.clamp(0.0, self.max_log_logit_scale).exp()
```
I don't see a diff with ClipModelFOrward!

yaswanth19 replied:
We actually don't have loss for Aimv2 and also there is a single LOC as we are adding `max_log_logit_scale` 
```suggestion
        logit_scale = self.logit_scale.clamp(0.0, self.max_log_logit_scale).exp()
```

ArthurZucker reviewed (approved):
Thanks ! A few small things to fix and good to go!

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Use more Modular 😉

yaswanth19 committed:
Change name from AIMv2 to Aimv2

yaswanth19 committed:
Nit

yaswanth19 commented:
@ArthurZucker  Addressed the Comments !!. IG you can transfer the checkpoints to Apple org and these one's have `HF` suffix. Later I will modify the checkpoint names in code and model card 🤗 

Checkpoints:
1.) `yaswanthgali/aimv2-large-patch14-224-lit-HF`
2.) `yaswanthgali/aimv2-large-patch14-native-HF`
3.) `yaswanthgali/aimv2-large-patch14-224-HF`

yaswanth19 commented:
One issue is I am seeing this warning `"You are using a model of type aimv2 to instantiate a model of type aimv2_vision_model. This is not supported for all configurations of models and can yield errors."` when loading ckpts related to image feature extarction. My hunch is we have `aimv2` as a base model prefix in `Aimv2PreTrainedModel` which is super class for `Aimv2VisionModel` and from there we are determining the type of model and comparing with the one in config.  I tried to overwrite base_model_prefix in `Aimv2VisionModel` and it didn't work 😢 .

yaswanth19 commented:
@ArthurZucker A ping for a few final things before merging

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

ArthurZucker commented:
On it ! Sorry 🤗

yaswanth19 commented:
@ArthurZucker  Yup NP 😄 
>Does it make sense to also convert most downloaded ones if we have to manually convert each one.

WDYT on the above statement? and also the issue which I have mentioned [above](https://github.com/huggingface/transformers/pull/36625#issuecomment-2842579859), even though it's harmless but it's a FP to the users. LMK how to resolve it 🙏

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
make style

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
@yonigozlan @ArthurZucker A ping for you attn here 🤗

yaswanth19 commented:
@ArthurZucker A gentle ping here. I know you have been busy with core stuff refactoring but lets try to wrap this up 🙏

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

ArthurZucker commented:
Yes, super sorry!

yaswanth19 committed:
Add model doc pointer

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
@ArthurZucker I missed adding model doc pointer in toctree because of which doc builder failed. Can you please re-run the workflow

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
make style

yaswanth19 committed:
Update model doc section

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
Hi @qubvel or maybe @Cyrilvallez — since Arthur is off, pinging you guys. The main things left are checkpoint transfer and updating it's references in the code. Would be great if we could get it merged. Thanks!

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,750 @@
+        return output
+
+
+AIMV2_VISION_START_DOCSTRING = r"""
+    This model inherits from [`PreTrainedModel`]. Check the superclass documentation for the generic methods the
+    library implements for all its model (such as downloading or saving, resizing the input embeddings, pruning heads
+    etc.)
+
+    This model is also a PyTorch [torch.nn.Module](https://pytorch.org/docs/stable/nn.html#torch.nn.Module) subclass.
+    Use it as a regular PyTorch Module and refer to the PyTorch documentation for all matter related to general usage
+    and behavior.
+
+    Parameters:
+        config ([`Aimv2VisionConfig`]): Model configuration class with all the parameters of the model.
+            Initializing with a config file does not load the weights associated with the model, only the
+            configuration. Check out the [`~PreTrainedModel.from_pretrained`] method to load the model weights.
+"""
+
+AIMV2_VISION_INPUTS_DOCSTRING = r"""
+    Args:
+        pixel_values (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)`):
+            Pixel values. Padding will be ignored by default should you provide it. Pixel values can be obtained using
+            [`AutoImageProcessor`]. See [`CLIPImageProcessor.__call__`] for details.
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        interpolate_pos_encoding (`bool`, *optional*, defaults `False`):
+            Whether to interpolate the pre-trained position encodings.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
```
We now use the `auto_docstring` decorator instead, so we can remove those variables

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,750 @@
+    config_class = Aimv2Config
+    base_model_prefix = "aimv2"
+    supports_gradient_checkpointing = True
+    _no_split_modules = ["AIMv2SwiGLUFFN"]
```
What is that module? Should be the EncoderLayer instead

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,750 @@
+        hidden_states = self.embeddings(input_ids)
+        _, seq_len, _ = hidden_states.shape
+
+        if attention_mask is not None:
+            mask_converter = AttentionMaskConverter(True)
+            attention_mask = mask_converter.to_4d(
+                attention_mask, key_value_length=seq_len, query_length=seq_len, dtype=hidden_states.dtype
+            )
```
We refactored the mask factory! It should be `create_causal_mask` instead, see Llama!

Cyrilvallez replied:
Ha indeed, you can use it without Cache but not without cache_position! I will look into maybe allowing this, this could indeed be helpful in general. In the meantime, it's best to still use `create_causal_mask`, simply use `cache_position = torch.arange(seq_len, device=hidden_state.device)` 🤗

Cyrilvallez commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -32,6 +32,8 @@
 MODEL_MAPPING_NAMES = OrderedDict(
     [
         # Base model mapping
+        ("aimv2", "Aimv2Model"),
+        ("aimv2_vision_model", "Aimv2VisionModel"),
```
Missing the text model part

Cyrilvallez replied:
Alright, if it aligns with public checkpoints it's fine! 🤗

Cyrilvallez commented on src/transformers/models/auto/configuration_auto.py:
```diff
@@ -36,6 +36,8 @@
 CONFIG_MAPPING_NAMES = OrderedDict[str, str](
     [
         # Add configs here
+        ("aimv2", "Aimv2Config"),
+        ("aimv2_vision_model", "Aimv2VisionConfig"),
```
missing text model part in the mappings

Cyrilvallez commented on tests/models/aimv2/test_modeling_aimv2.py:
```diff
@@ -0,0 +1,742 @@
+# coding=utf-8
+# Copyright 2021 The HuggingFace Inc. team. All rights reserved.
```
2025!

Cyrilvallez commented on utils/check_repo.py:
```diff
@@ -195,6 +195,7 @@
 # should **not** be the rule.
 IGNORE_NON_AUTO_CONFIGURED = PRIVATE_MODELS.copy() + [
     # models to ignore for model xxx mapping
+    "Aimv2TextModel",
```
We can remove this and add it to the auto mappings instead as it's tested 👌

yaswanth19 replied:
Same comment as above

Cyrilvallez reviewed (commented):
Hey! Super sorry for the delay! Here are a few new comments based on more recent refactors 🤗 But it's very good and almost ready to be merged!

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
updates

Cyrilvallez reviewed (approved):
ALright! Final comment here, + the one above about the mask and it's ready to go! Great work! 🤗

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Modify attn mask and interface

yaswanth19 committed:
update test

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 commented:
@Cyrilvallez Waiting for PR #39040 to get merged which will help in modular for this PR. It will also resolve the CI and then you can merge this PR. Feel free to review that PR incase Arthur doesn't have the bandwidth :hugs:

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Final change

Cyrilvallez committed:
Merge branch 'main' into add-aimv2-model

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,724 @@
+        self.max_log_logit_scale = math.log(config.max_logit_scale)
+
+        self.post_init()
+
+    def forward(
```
Missing `auto_docstring` and `can_return_tuple` here

yaswanth19 replied:
It was present in modeling file coz ig we were just overwriting forward, yeah but no loss in being explicit in modular.

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,724 @@
+
+        hidden_states = self.embeddings(input_ids)
+        _, seq_len, _ = hidden_states.shape
+
+        if attention_mask is not None:
```
Here we should remove the check altogether (I think it was suppoosed to check the opposite anyway, but `create_causal_mask` will work independently of the mask being None or not)

yaswanth19 replied:
@Cyrilvallez  Not quite working correctly, in eager implementation we are always returning a mask even when attention_mask is None whereas in sdpa implementation we are correctly returning None. AFAIK in eager mode we are creating the mask coz `allow_is_causal_skip` is False by default (so sdpa func returns a causal mask in this but in sdpa implementation that flag is True) :thinking: but tbh didn't completely get the new mask utils logic so I might wrong in my understanding.

Cyrilvallez replied:
Ha yes it will always return a mask because it creates a causal one! If the attention is not causal and we never use a mask, then we should remove that function altogether!

Cyrilvallez commented on src/transformers/models/aimv2/modular_aimv2.py:
```diff
@@ -0,0 +1,724 @@
+        "Aimv2VisionEmbeddings",
+        "Aimv2TextEmbeddings",
+    ]
+    _supports_sdpa = True
```
We should support FA2 and flex as well!

Cyrilvallez reviewed (approved):
This time it's the final nits I promise! 🤗🤪

yaswanth19 committed:
Utilize flash and flex attn

yaswanth19 committed:
keep attn mask

yaswanth19 committed:
camelcase model name in test file

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Merge branch 'main' into add-aimv2-model

yaswanth19 committed:
Fix docstring

yaswanth19 committed:
Fix config warning finally and create_causal_mask

yaswanth19 committed:
disable torchscript

Cyrilvallez committed:
remove unused arg

Cyrilvallez committed:
Merge branch 'main' into add-aimv2-model

Cyrilvallez committed:
remove from tests

Cyrilvallez committed:
balance model size for tests

Cyrilvallez committed:
fix device

Cyrilvallez committed:
tests

Cyrilvallez committed:
tests

Cyrilvallez commented:
run-slow: aimv2

Cyrilvallez committed:
flaky test

Cyrilvallez committed:
fix import

Cyrilvallez reviewed (approved):
All right! I pushed the last few fixes myself as you saw, this is not ready to be merged! 
Thanks again for the great work and contribution! 🤗🚀 And thanks for bearing with us during the review process!

nahidalam commented:
Which version of Huggingface is AIMv2 added? Or it is still in dev?
